### ISSUE TITLE
screencast: dmabuf support

### ISSUE THREAD

**Issue by danshick** [emersion/xdg-desktop-portal-wlr#9]
Support passing a dmabuf fd to pipewire instead of using the screencopy protocol. There are three things pending before this can (or should) move forward.

1. ~~Waiting on Pipewire 0.3 to be released:
    https://gitlab.freedesktop.org/pipewire/pipewire~~

2. ~~Waiting on xdg-desktop-portal PR to support Pipewire 0.3:
    https://github.com/flatpak/xdg-desktop-portal/pull/436~~

3. ~~Waiting on downstream ecosystem to support drm_fourcc modifiers:
    [Discussion of downstream dependencies](https://github.com/emersion/xdg-desktop-portal-wlr/issues/1#issuecomment-571684093) that need to support modifiers.~~ resolved by newer screencopy protocol

[Here](https://github.com/danshick/wlr_zerocopy_pipewire) is a very early, broken demo

**Comment by braderhart** [comment]
Pipewire 0.3 is out now. Hope that will help clear up the issues.

**Comment by danshick** [comment]
That certainly means we can complete dmabuf style screensharing, but it'll be mostly useless unless and until downstream apps support various modifiers. Still worth building out, although it'll likely get hidden behind an argument.

**Comment by danshick** [comment]
Maybe we could use glBlitFramebuffer to copy frames on GPU before sharing. This could avoid the concern about format modifiers. I need to research and experiment with this approach further.

**Comment by emersion** [comment]
https://github.com/swaywm/wlroots/pull/2133 allows blitting DMA-BUFs on the compositor-side.

**Comment by emersion** [comment]
>Do you think this should be implemented under a CLI flag?

Ideally it should be enabled by default. We should offer support for both DMA-BUF and good old shm, and clients should pick one of those.

**Comment by danshick** [comment]
I totally forgot that Mutter already does this (dmabuf producer for xdg-desktop-portal-gtk, consumed by WebRTC). Basically, this should work fine in all cases.

We can't let clients pick one because there is no provision in the portal API for that negotiation to happen, but luckily, I don't think we have to.

The only fallback that would be necessary is if the screencopy protocol version is insufficient on the compositor side.

**Comment by emersion** [comment]
FWIW, https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1442

I think GNOME also blits to LINEAR because all clients ignore the modifier hint and end up with garbage.

Pipewire is really missing a better API for this. Giving DMA-BUFs to a client expecting shm is not going to end well.

**Comment by columbarius** [comment]
I guess currently dmabuf support isn't fully implemented in pipewire at the moment. I guess if a client ca't work with a fd, pipewire will copy it to it's data pointer itself: https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/stream.c#L548
It's done the same way for memfd and dmabuf without awareness of modifiers. I guess creating a separate path using gbm would  probably solve some of the issues. Will ask at #pipewire for confirmation.

**Comment by danshick** [comment]
> I guess if a client ca't work with a fd, pipewire will copy it to it's data pointer itself: https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/stream.c#L548

Pipewire always maps buffers for the client. It isn't a question of client support. The good news is that mutter being broken will definitely help to make the case for a better dmabuf handling story.

**Comment by danshick** [comment]
I'll need to look into the code regarding the dataType masks. It should be trivial to use the dmabuf extension to screencopy right now if we can rely on clients to handle it correctly.

As for export dmabuf, I'm guessing there could still be timing issues there if some downstream client is holding a frame fd for too long? I know that was discussed a while back. Was there ever a decision made about how wlroots handles those cases?

**Comment by columbarius** [comment]
Thanks. I guess we have to revert https://github.com/emersion/xdg-desktop-portal-wlr/commit/c2883a4e43aca02545151a894503fdb6c7da58c4. 

Does ~0 means NULL?

I guess we need some logic [here](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/src/screencast/pipewire_screencast.c#L114) to select the correct buffer type/backend.

**Comment by danshick** [comment]
`~` is the urinary operator for the bitwise compliment, so ~0 would be a binary value of all ones, which varies in length and value depending on the type.

**Comment by emersion** [comment]
>As for export dmabuf, I'm guessing there could still be timing issues there if some downstream client is holding a frame fd for too long? I know that was discussed a while back. Was there ever a decision made about how wlroots handles those cases?

This is a theoretical issue, yes. I'd say let's just stick to screencopy for now and leave export-dmabuf for later.

**Comment by columbarius** [comment]
Some update to dmabufs in pipewire:

Exampleclients to test format merging: 
https://gitlab.freedesktop.org/columbarius/pipewire-modifier-test

Obs branch to have a client to test on (WIP):
https://github.com/columbarius/obs-studio/tree/egl-modifiers

Found the funktion in pipewire responsible for merging the formats:
https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/src/pipewire/context.c#L724
The format_filter doesn't look used or implemented, maybe that's sth. work on?

**Comment by columbarius** [comment]
When this is merged, we can reliably distinguish between dmabufs and MemPtr/MemFd

https://gitlab.freedesktop.org/pipewire/pipewire/-/merge_requests/683

**Comment by columbarius** [comment]
Update: 
https://gitlab.freedesktop.org/pipewire/pipewire/-/merge_requests/683
got merged into pipewire 0.3.29

**Comment by columbarius** [comment]
PipeWire 0.3.40 was released with full support for negotiating and dropping single modifiers!
